### ISSUE TITLE
NSE case studies

### ISSUE THREAD

**Issue by soto-13** [zongyi-li/fourier_neural_operator#28]
Hello,

I was interested in the NSE implementation (fourier_2d_time.py and fourier_3d.py).  I used input data NavierStokes_V1e-5_N1200_T20.zip (from the Drive), and I understand it is structured as: [number of samples x image height x image width x timestep] 

Trying it out I got the following issues:

**fourier_2d_time.py** (uncommenting the lines to save the model and evaluate (not sure why it was commented)):
- `out = y_normalizer.decode(out) (l. 328)` shows an undefined normalizer, I assume this is just an extra line because no normalization is done in this file (as opposed to fourier_3d.py where it is done).
- `test_l2 += myloss(out.view(1, -1), y.view(1, -1)).item()` shows tensor size inconsistencies (model output `out` has 1 time-step versus 10 for `y`)

**fourier_3d.py**: (edit - ok)

Given that both loss function evaluations are fairly standard I might have omitted to adapt something to this specific case (edit: `out` uses single time-step output while `y` has the whole information - hence the error), the rest are minor fixes

Best wishes

**Comment by zongyi-li** [comment]
Sorry the commented code in the end is a bit outdated. (1) Yes, the normalizer is not needed. (2) please copy the test code for the validation (line 287-310) to do the testing. For this time series version, it needs an additional for-loop in time. The input is the first 10 timesteps, and for each iteration, it should predict the next time step.

**Comment by zongyi-li** [comment]
For your information, the dataset NavierStokes_V1e-5_N1200_T20.zip meaning viscosity v = 1e-5, which is the most challenging case. It's better to start with the V1e-3 dataset.

**Comment by soto-13** [comment]
Ah ok I managed to run the V1e-3 on `fourier_3d.py`. Is it normal though that:
```
train_a = reader.read_field('u')[:ntrain, ::sub, ::sub, :T_in]
train_u = reader.read_field('u')[:ntrain, ::sub, ::sub, T_in:T_in+T]
```
(l. 223-224, l. 227-228) i.e. I would expect read_field('a') for `train_a` for instance. Matlab shows that the dataset is split into the fields `a`, `u`, `t` hence my concern

**Comment by zongyi-li** [comment]
`a` is the initial time step. `u` is the full 50 time steps. In this setting, we are giving the first 10 time steps to predict the next 40 time steps, so both the input and output are read from `u`. Sorry for the confusion.

**Comment by kaushikb258** [comment]
Hi, nice work. I am trying to simulate the NS problem with fourier_2d_time.py using ns_V1e-4_N10000_T30.mat data. I want to generate figures similar to Fig. 1b in the paper. For this do I plot the variables yy and pred (from the bottom of fourier_2d_time.py)? 

Also, how to choose modes and width in fourier_2d_time.py? I  use modes = 24 and width = 64. You have modes = 12 and width = 20 as default. Do these values change?

**Comment by zongyi-li** [comment]
Thanks. Yes, to get the test case as 1(b) you should use the code at the bottom.  But I think this code for fourier_2d_time.py was outdated. You need to copy the validation code in the training period (line 287-231). Yes, you should plot variables yy and pred. 

Super-resolution is easier on fourier_3d.py

**Comment by kaushikb258** [comment]
Hi, I tried both fourier_2d_time and fourier_3d on the Navier-Stokes dataset. However, the predictions are very blurred compared to the ground truth. I am trying to generate results similar to Fig. 1b in the paper. For this, can you please suggest if the default hyperparams in the code (e.g. modes, width, epochs, etc) are set for the Navier-Stokes problem or should they be set to other values. What values were used to generate Fig. 1b results? Thanks!

**Comment by zongyi-li** [comment]
for fourier_2d_time and the v1e-4, N10000 datatset, please use modes=12, width=32, epochs=200, with the learning rate decay by half for every 40 epochs. fourier_3d, please use modes=8, width=20 instead.
